kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ReverseAddRemoveToAddColumnarBatchIterator.java

```diff
@@ -0,0 +1,235 @@
+            }
+        }
+
+        // Step 2: Iterate over all the AddFiles in this columnar batch in order to build up the
+        //         selection vector. We unselect an AddFile when it was removed by a RemoveFile
+        final ColumnVector addsVector = addRemoveColumnarBatch.getColumnVector(0);
+        boolean[] selectionVector = new boolean[addsVector.getSize()];
+        boolean atLeastOneUnselected = false;
+
+        for (int rowId = 0; rowId < addsVector.getSize(); rowId++) {
+            if (addsVector.isNullAt(rowId)) {
+                atLeastOneUnselected = true;
+                continue; // selectionVector will be `false` at rowId by default
+            }
+
+            final Row row = addsVector.getStruct(rowId);
+            final String path = AddFile.getPathFromRow(row);
+            final URI pathAsUri = pathToUri(path);
+            final Optional<String> dvId = Optional.ofNullable(
+                AddFile.getDeletionVectorDescriptorFromRow(row)
+            ).map(DeletionVectorDescriptor::getUniqueId);
+            final UniqueFileActionTuple key = new UniqueFileActionTuple(pathAsUri, dvId);
+            final boolean alreadyDeleted = tombstonesFromJson.contains(key);
+            final boolean alreadyReturned = addFilesFromJson.contains(key);
+
+            boolean doSelect = false;
+
+            if (!alreadyReturned) {
+                // Note: No AddFile will appear twice in a checkpoint, so we only need
+                //       non-checkpoint AddFiles in the set
+                if (!isFromCheckpoint) {
+                    addFilesFromJson.add(key);
+                }
+
+                if (!alreadyDeleted) {
+                    doSelect = true;
+                    selectionVector[rowId] = true;
+                }
+            }
+
+            if (!doSelect) {
+                atLeastOneUnselected = true;
+            }
+        }
+
+        // Step 3: Drop the RemoveFile column and use the selection vector to build a new
+        //         DataReadResult
+        next = Optional.of(
+            new DataReadResult(
+                addRemoveColumnarBatch.withDeletedColumnAt(1),
+                atLeastOneUnselected ?
+                    Optional.of(boolArrayToColumnVector(selectionVector)) :
+                    Optional.empty()
+            )
+        );
+    }
+
+    private ColumnVector boolArrayToColumnVector(boolean[] arr) {
+        return new ColumnVector() {
+            @Override
+            public DataType getDataType() {
+                return BooleanType.INSTANCE;
+            }
+
+            @Override
+            public int getSize() {
+                return arr.length;
+            }
+
+            @Override
+            public void close() { }
+
+            @Override
+            public boolean isNullAt(int rowId) {
+                return false;
+            }
+
+            @Override
+            public boolean getBoolean(int rowId) {
+                return arr[rowId];
```

Comment:
valid `rowId` check

Reply:
Doesn't java do this for us, anyways, on every array access? Or do you want us to throw a nice error?

Reply:
yeah, it throws ArrayOutOfBoundException, but we don't where it happened. Check the DefaultColumnVector implementations in default module.